### PR TITLE
Implement ScheduledTask Pattern for FastHotStuffEventLoop Timers

### PR DESCRIPTION
**Pull request by tholonious**

### CODE REVIEW

**Comment by tholonious** [comment]
Current dependencies on/for this PR:
* feature/proof-of-stake
  * **PR #649** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/649" target="_blank">Graphite</a> 
    * **PR #648** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/648" target="_blank">Graphite</a> 
      * **PR #650** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/650" target="_blank">Graphite</a> 
        * **PR #654** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/654" target="_blank">Graphite</a> 
          * **PR #656** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/656" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/deso-protocol/core/656?utm_source=stack-comment).

File: consensus/scheduled_task.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
+package consensus
+
+import (
+	"sync"
+	"time"
+)
+
+// ScheduledTask is a thread-safe wrapper around time.Timer that allows for creating tasks that
+// can be scheduled to execute at a later time with pre-specified params. Both the params the
```
Comment (lazynina) [inline consensus/scheduled_task.go:9]:
```suggestion
// can be scheduled to execute at a later time with pre-specified params. Both the params and the
```

File: consensus/scheduled_task.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+func NewScheduledTask[TaskParam any]() *ScheduledTask[TaskParam] {
+	return &ScheduledTask[TaskParam]{
+		lock:  sync.RWMutex{},
+		timer: nil,
+	}
+}
+
+// Cancel the currently scheduled task, if any, and schedule a new task to be executed after the countdown duration.
```
Comment (lazynina) [inline consensus/scheduled_task.go:28]: should this comment be next to the cancel function?

File: consensus/scheduled_task_test.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+	// Test short scheduled task
+	{
+		task := NewScheduledTask[uint64]()
+
+		executedTaskParam := uint64(0)
+		task.Schedule(time.Microsecond, 100, func(param uint64) {
+			executedTaskParam = param
+		})
+		time.Sleep(time.Second)
```
Comment (lazynina) [inline consensus/scheduled_task_test.go:20]: Should we reduce the sleep to something shorter?

File: consensus/scheduled_task_test.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+		// The task should not have executed so this value will now be 100.
```
Comment (lazynina) [inline consensus/scheduled_task_test.go:22]:
```suggestion
		// The task should have executed so this value will now be 100.
```

File: consensus/scheduled_task_test.go [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+	// Test long scheduled task that's not expected to execute
+	{
+		task := NewScheduledTask[uint64]()
+
+		executedTaskParam := uint64(0)
+		task.Schedule(time.Hour, 100, func(param uint64) {
+			executedTaskParam = param
+		})
+		time.Sleep(time.Second)
```
Comment (lazynina) [inline consensus/scheduled_task_test.go:37]: same here - reduce sleep?

**Review by lazynina — CHANGES_REQUESTED** [review]

File: consensus/scheduled_task.go (hunk 2 above)
Comment (tholonious) [reply consensus/scheduled_task.go:28]: Nah, this is the correct place. I've reworded the comment so it's a bit cleraer

File: consensus/scheduled_task_test.go (hunk 5 above)
Comment (tholonious) [reply consensus/scheduled_task_test.go:37]: We can get rid of the sleep here

File: consensus/scheduled_task_test.go (hunk 3 above)
Comment (lazynina) [reply consensus/scheduled_task_test.go:20]: I'm down w/ 0.5 seconds.

Comment (tholonious) [reply consensus/scheduled_task_test.go:20]: Sounds good

**Review by lazynina — APPROVED** [review]

**Pull request merged**